File: lib/models/bill_adjustment.dart
import 'package:equatable/equatable.dart';

class BillAdjustment extends Equatable {
  final String id;
  final String billId;
  final String customerId;
  final double amount;
  final String reason;
  final String note;
  final DateTime adjustmentDate;
  final DateTime? createdAt;

  const BillAdjustment({
    required this.id,
    required this.billId,
    required this.customerId,
    required this.amount,
    required this.reason,
    this.note = '',
    required this.adjustmentDate,
    this.createdAt,
  });

  factory BillAdjustment.fromJson(Map<String, dynamic> json) => BillAdjustment(
    id: json['_id'] ?? json['id'] ?? '',
    billId: json['billId'] ?? '',
    customerId: json['customerId'] ?? '',
    amount: (json['amount'] ?? 0).toDouble(),
    reason: json['reason'] ?? 'other',
    note: json['note'] ?? '',
    adjustmentDate: DateTime.parse(json['adjustmentDate'] ?? DateTime.now().toIso8601String()),
    createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
  );

  String get reasonLabel {
    switch (reason) {
      case 'damaged':
        return 'Damaged';
      case 'missing':
        return 'Missing';
      case 'short_supply':
        return 'Short Supply';
      case 'rate_diff':
        return 'Rate Difference';
      default:
        return 'Other';
    }
  }

  @override
  List<Object?> get props => [id, billId, amount, reason, note];
}

File: backend/src/models/BillAdjustment.js
const mongoose = require('mongoose');

const billAdjustmentSchema = new mongoose.Schema({
  billId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Bill',
    required: true,
    index: true,
  },
  customerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Customer',
    required: true,
  },
  billItemId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'BillItem',
    default: null,
    index: true,
  },
  originalQuantity: {
    type: Number,
    default: null,
  },
  adjustedQuantity: {
    type: Number,
    default: null,
  },
  amount: {
    type: Number,
    required: true,
    min: 0.01,
  },
  reason: {
    type: String,
    required: true,
    enum: ['damaged', 'missing', 'short_supply', 'rate_diff', 'other'],
  },
  note: {
    type: String,
    default: '',
    maxlength: 300,
  },
  adjustmentDate: {
    type: Date,
    default: Date.now,
    required: true,
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  },
}, { timestamps: true });

billAdjustmentSchema.index({ billId: 1, createdAt: -1 });
billAdjustmentSchema.index({ customerId: 1 });

module.exports = mongoose.model('BillAdjustment', billAdjustmentSchema);
